src/main/java/org/opentripplanner/model/plan/legreference/ScheduledTransitLegReference.java

```diff
@@ -123,4 +143,60 @@ public ScheduledTransitLeg getLeg(TransitService transitService) {
 
     return leg;
   }
+
+  /**
+   * Return false if the stop id in the reference does not match the actual stop id in the trip
+   * pattern.
+   * Return true in the specific case where the stop ids differ, but belong to the same parent
+   * station.
+   *
+   */
+  private boolean matchReferencedStopInPattern(
+    TripPattern tripPattern,
+    int stopPosition,
+    FeedScopedId stopId,
+    TransitService transitService
+  ) {
+    if (stopId == null) {
+      // this is a legacy reference, skip validation
+      // TODO: remove backward-compatible logic after OTP release 2.6
+      return true;
+    }
+
+    StopLocation stopLocationInPattern = tripPattern.getStops().get(stopPosition);
+    if (stopId.equals(stopLocationInPattern.getId())) {
+      return true;
+    }
+    StopLocation stopLocationInLegReference = transitService.getStopLocation(stopId);
+    if (
+      stopLocationInLegReference == null ||
+      stopLocationInPattern.getParentStation() == null ||
+      !stopLocationInPattern
+        .getParentStation()
+        .equals(stopLocationInLegReference.getParentStation())
+    ) {
+      LOG.info(
```

Comment:
Isn't info log level a bit noisy for normal operations?

Reply:
These are corner-cases, not expected to be numerous, and it is important to get an idea of how many invalid legs are received in production (this is more than debug information).